### PR TITLE
feat(components): implement Highlight agnostic core

### LINKED ISSUE #204: task: Implement Highlight agnostic core
**Epic: #10 (Agnostic utility components)**

## Goal

Implement the framework-agnostic Highlight component: text substring matching with three strategies (Contains, StartsWith, Fuzzy), Unicode-aware case folding via ICU4X `CaseMapper`, multi-query support with overlapping range merging, and `HighlightChunk` output.

## Out of scope

Framework-specific adapter components (Leptos/Dioxus).

## Spec refs

- `spec/components/utility/highlight.md`

## Depends on

None

## Layer

Component

## Framework

None

## Test tier

Unit

## Fibonacci points

3

## Tests to add first

- Contains strategy: matches substring and produces correct HighlightChunk spans.
- StartsWith strategy: matches only at word/string boundaries.
- Fuzzy strategy: matches characters in order with gaps.
- Unicode case folding: Turkic locale (İ→i), German eszett (ß→ss) handled correctly.
- Multi-query: overlapping ranges merged into single highlighted span.
- Empty query: returns single non-highlighted chunk.
- Props: `ignore_case` flag toggles case-sensitive matching.

## Acceptance criteria

- All public types documented per workspace `missing_docs` lint.
- Three match strategies implemented per spec.
- Unicode case folding uses ICU4X CaseMapper, not ASCII-only lowering.
- HighlightChunk output matches spec API exactly.
- Snapshot tests cover the component's `connect()` / `Api` `AttrMap` output for every anatomy part and each output-affecting state, prop, or context branch, with committed `.snap` fixtures passing `cargo insta test --unreferenced=reject`.
- All named tests pass with zero warnings.

## Spec impact

No spec change required.

## Board update rule

If blocked, keep `Status` in `Todo`, use native GitHub issue dependencies, and name the dependency or spec ambiguity.

### PR DESCRIPTION
**Pull request by fogodev**
## Summary

- Implement framework-agnostic `Highlight` utility (`ars-components::utility::highlight`) per `spec/components/utility/highlight.md` — three match strategies (Contains/StartsWith/Fuzzy), TR21 Unicode case folding, multi-query overlap+adjacency merging, parametric chunk anatomy. Gated on a new `i18n` feature wired through the adapter `icu4x` features so Leptos / Dioxus builds with icu4x pick up Highlight automatically.
- Surface a new `ars_i18n::case_fold` helper that wraps `CaseMapper::fold_string` / `fold_turkic_string`. Switching from `to_lowercase` closes the spec's eszett contract — `"ss"` now genuinely matches `"ß"` in both directions (the original implementation passed its named tests but quietly violated this).
- Codify the audit flow that drove this PR: new `post-implementation-audit` skill at `.agents/skills/post-implementation-audit/SKILL.md`, registered as mandatory step 7 of the AGENTS.md / CLAUDE.md development workflow. Three sequential phases (spec/impl drift, iterative "anything missing?", test coverage) that land every finding in the same PR. The skill's worked example references this PR.

Closes #204.

## Test plan

- [x] `cargo xci` — all 20 steps pass (fmt, check, clippy, unit, i18n-browser, dom-browser, release, integration, adapter, adapter-parity, coverage, snapshot-count, spec-compile-snippets, error-variant-coverage, 5 feature matrices, mutual-exclusion)
- [x] `cargo test -p ars-components --features i18n --all-targets` — 1502 lib + 4 conformance + 1 doc tests pass
- [x] `cargo test -p ars-components --lib` (no `i18n`) — 1463 pass; Highlight module gated out cleanly
- [x] `cargo test -p ars-i18n --features icu4x case::` — 10 pass (incl. 4 new `case_fold` tests covering eszett expansion, Greek final sigma collapse, Turkic vs non-Turkic, idempotence)
- [x] `cargo mutants -p ars-components -f crates/ars-components/src/utility/highlight.rs` — **57/57 caught, 0 missed** (3 equivalent mutations documented with rationale in `.cargo/mutants.toml`)
- [x] `cargo llvm-cov` — **99.91% line / 100% function / 100% region** for `utility/highlight.rs`
- [x] `cargo xtask spec validate` — 340 files clean
- [x] `cargo build -p ars-leptos --features icu4x` and `cargo build -p ars-dioxus --features icu4x,web` — Highlight reachable in adapter builds
- [x] Spec-conformance test asserts the `Part` enum matches the spec's §2 anatomy table

## What landed

| Area | Files |
|---|---|
| Agnostic core | `crates/ars-components/src/utility/highlight.rs` (new), `crates/ars-components/src/utility/mod.rs`, `crates/ars-components/Cargo.toml` |
| Locale-aware case folding helper | `crates/ars-i18n/src/case.rs`, `crates/ars-i18n/src/lib.rs` |
| Adapter feature wiring | `crates/ars-leptos/Cargo.toml`, `crates/ars-dioxus/Cargo.toml` |
| Tests | 9 insta snapshots + inline doc test + 1 spec-conformance test + 1 proptest invariant block + 41 unit tests |
| Spec | Core spec rewrite, Leptos/Dioxus adapter spec updates, foundation/10 new §4.2 parametric-anatomy convention |
| Workflow codification | `.agents/skills/post-implementation-audit/SKILL.md`, `AGENTS.md` step 7 mandatory invocation, `.cargo/mutants.toml` skip justifications |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/ars-components/src/utility/highlight.rs [hunk 1]
```diff
@@ -0,0 +1,1346 @@
...
+            case_fold(ch_str, locale)
+        } else {
+            let mut s = String::new();
+
+            s.push(ch);
+
+            s
+        };
+
+        if normalised_ch_string.chars().any(|c| c == want) {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/highlight.rs]: Fixed in f670b8ef.

`push_fuzzy_matches` now consumes **every** leading query character that appears, in order, inside each source character's folded expansion — not just the first. So `ß`'s `ss` fold under German absorbs both query `s`s in one source-char iteration, the query iterator empties out, and the all-or-nothing rule lets the fuzzy match succeed. New test `fuzzy_consumes_full_folded_expansion_for_eszett` pins this exact case; the existing `chunks_concatenate_to_original_text_across_strategies` matrix covers the inverse and asymmetric combos. Mutation testing still clean (55/55 caught).

File: crates/ars-components/src/utility/highlight.rs [hunk 2]
```diff
@@ -0,0 +1,1346 @@
...
+
+use ars_core::{AttrMap, ComponentPart, ConnectApi, HtmlAttr};
+use ars_i18n::Locale;
+
+/// Props for the [`Highlight`](crate::utility::highlight) component.
+///
+/// `Highlight` does not carry a DOM `id` — chunk output is a sequence of
+/// adapter-rendered children, not a single identifiable element — so [`Props`]
+/// does not derive [`HasId`](ars_core::HasId) like other stateless utilities.
+#[derive(Clone, Debug, PartialEq)]
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/highlight.rs]: Fixed in f670b8ef.

The derive list on `Props` is now `#[derive(Clone, Debug, PartialEq, Eq)]`. `Default` stays as an explicit `impl Default for Props` rather than going into the derive — the spec's non-field-default contracts (`ignore_case = true`, `match_strategy = MatchStrategy::Contains`) can't be expressed by `#[derive(Default)]`, which would set everything to the field type's own default. Updated `spec/components/utility/highlight.md` §1.1 to match (the derive list with a doc comment explaining the manual `impl Default`).

File: crates/ars-components/src/utility/highlight.rs [hunk 3]
```diff
@@ -0,0 +1,1346 @@
...
+    ars_i18n::case_fold(text, locale)
+}
+
+fn push_contains_matches(
+    normalised_text: &str,
+    normalised_query: &str,
+    byte_map: &[usize],
+    out: &mut Vec<(usize, usize)>,
+) {
+    for (start, _) in normalised_text.match_indices(normalised_query) {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/highlight.rs]: Fixed in f670b8ef.

`push_contains_matches` no longer uses `str::match_indices` (which yields only non-overlapping matches). It now scans with `find` and advances by one **character** past each match's start, so overlapping occurrences surface and then collapse via `merge_ranges`'s adjacency/overlap merge into a single highlighted span. New test `contains_strategy_finds_overlapping_occurrences` pins the `"ana"`/`"banana"` case you flagged — it produces matches at byte offsets 1 and 3, merges into the single `"anana"` highlight. `spec/components/utility/highlight.md` §1.1 also rewords `MatchStrategy::Contains` to call out the overlap behaviour explicitly.

File: crates/ars-components/tests/proptest_state_machines/utility.rs [hunk 4]
```diff
@@ -694,3 +698,131 @@ proptest! {
...
+        Just(Locale::parse("de").expect("de must parse")),
+        Just(Locale::parse("el").expect("el must parse")),
+        Just(Locale::parse("lt").expect("lt must parse")),
+        Just(Locale::parse("hy").expect("hy must parse")),
+        Just(Locale::parse("ar").expect("ar must parse")),
+    ]
+}
+
+#[cfg(feature = "i18n")]
+proptest! {
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/tests/proptest_state_machines/utility.rs:768]: Fixed in f670b8ef.

`.github/workflows/nightly.yml` now runs `cargo test -p ars-core -p ars-components --features ars-components/i18n -- --ignored proptest`, so the feature-gated Highlight invariant proptest (roundtrip, no-adjacent-highlighted, no-empty-chunks, empty-edge-cases × 8 locales) actually compiles and runs in the nightly extended-proptest job. Without the feature flag the test silently disappeared from the suite even though the binary compiled — exactly the failure mode you described.

File: crates/ars-components/src/utility/highlight.rs [hunk 5]
```diff
@@ -0,0 +1,1346 @@
...
+
+    for (orig_byte, ch) in text.char_indices() {
+        let mut buf = [0u8; 4];
+
+        let ch_str = ch.encode_utf8(&mut buf);
+
+        let folded = case_fold(ch_str, locale);
+
+        for _ in 0..folded.len() {
+            byte_map.push(orig_byte);
```
[the remark this answers is not in the document]
Comment (fogodev) [reply crates/ars-components/src/utility/highlight.rs]: Fixed in f670b8ef.

`Normalised::build` now records two parallel arrays — `source_start[i]` (start byte of the source char that produced folded byte `i`) and `source_end[i]` (its end byte). `Normalised::map_range(a, b)` returns `(source_start[a], source_end[b - 1])`, so a match ending inside a fold expansion extends to cover the **full** contributing source character.

For the cases you described:

- `StartsWith` with `query=["stras"]` against German `text="Straße"` — folded text is `strasse`, match `[0, 5)`. `source_end[4] = 6` (end of `ß`), so the highlighted span is `Straß`, not `Stra`. New test `starts_with_extends_match_to_full_fold_expansion_source_char` pins this.
- `query=["s"]` against `text="ß"` — match `[0, 1)`. `source_end[0] = 2` (end of `ß`), so the whole `ß` highlights instead of being dropped as a zero-length range. Covered transitively by the new test plus `fold_expansion_partial_match_highlights_full_source_char` (which exercises `İ`/`i\u{307}` for both Contains and StartsWith).

The old "zero-length range filter" guards in `push_contains_matches` (`orig_start < orig_end`) and `push_starts_with_match` (`orig_end > 0`) are gone — they were filtering out exactly the matches that should now produce a source-char-bounded highlight. Mutation testing recalibrated: 55/55 caught, with the now-unreachable filter excludes removed from `.cargo/mutants.toml`.
